### PR TITLE
DIV-4679 - Case not found mapping

### PR DESCRIPTION
**Pull request by hosainnet**
# Description

* Map case not found errors from CCD (400) to 401 to remove bad request noise from alerts and make it consistent for the FE to handle
* Added the ability to run integration tests locally in an easier way
* Attempt to make functional tests less flakey by closing HTTP connection

Fixes #DIV-4679

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,3 +20,6 @@ target/
 
 src/main/resources/version
 java-rpm-packaging/
+application-*.yml
+application-*.yaml
+application-local.properties
```
Comment (hosainnet) [inline .gitignore]: Ignore files with senestive info

File: build.gradle [hunk 2]
```diff
@@ -277,6 +279,15 @@ task smoke(type: Test, description: 'Runs the smoke tests.', group: 'Verificatio
...
 
+task bootRunAat(type: BootRun, description: 'Runs the app using AAT config', dependsOn: 'classes', group: 'Application') {
+    environment "http_proxy", "http://proxyout.reform.hmcts.net:8080"
+    doFirst() {
+        main = bootJar.mainClassName
+        classpath = sourceSets.main.runtimeClasspath
+        systemProperty 'spring.profiles.active', 'aat'
+    }
+}
+
```
Comment (hosainnet) [inline build.gradle]: Run the app locally in AAT mode

File: charts/div-cms/requirements.yaml [hunk 3]
```diff
@@ -1,9 +1,4 @@
 dependencies:
     - name: java
-      version: ~1.1.2
+      version: ~2.0.4
       repository: "@hmcts"
-    - name: ccd
-      version: ~0.6.0
-      repository: '@hmcts'
-      tags:
-          - ccd
```
Comment (hosainnet) [inline charts/div-cms/requirements.yaml]: Removed for now, will add soon

File: charts/div-cms/values.preview.template.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        CASE_FORMATTER_SERVICE_API_BASEURL : "http://div-cfs-aat.service.core-compute-aat.internal"
+        REFORM_ENVIRONMENT : "aat"
+        IDAM_API_BASEURL: "https://idam-api.aat.platform.hmcts.net"
+        IDAM_API_REDIRECT_URL : "https://div-pfe-aat.service.core-compute-aat.internal/authenticated"
+        CASE_DATA_STORE_BASEURL : "http://ccd-data-store-api-aat.service.core-compute-aat.internal"
+        APPINSIGHTS_INSTRUMENTATIONKEY : "63c30ef9-a27b-4d38-8afc-4c180fb6ac33"
+
+    # Don't modify below here
+    image: ${IMAGE_NAME}
+    ingressHost: ${SERVICE_FQDN}
```
Comment (hosainnet) [inline charts/div-cms/values.preview.template.yaml]: Added a specific preview env file as per instructions from Helm team

File: charts/div-cms/values.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+        AUTH_PROVIDER_SERVICE_CLIENT_TOKENTIMETOLIVEINSECONDS: "900"
+        LOG_LEVEL: DEBUG
+    keyVaults:
+        "div":
+            resourceGroup: div
+            secrets:
+                - ccd-submission-s2s-auth-secret
+                - idam-secret
+                - draft-store-api-encryption-key
+                - idam-caseworker-password
```
Comment (hosainnet) [inline charts/div-cms/values.yaml]: Renamed from `values.template.yaml`

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java [hunk 6]
```diff
@@ -26,6 +37,23 @@ protected IntegrationTest() {
...
+                System.setProperty("http.proxyPort", Integer.toString(proxy.getPort()));
+                System.setProperty("https.proxyHost", proxy.getHost());
+                System.setProperty("https.proxyPort", Integer.toString(proxy.getPort()));
+            } catch (IOException e) {
+                log.error("Error setting up proxy - are you connected to the VPN?", e);
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
```
Comment (hosainnet) [inline src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java]: Set proxy if local properties has it

File: src/integrationTest/resources/example-application-local.properties [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+# Hardcoded for Kubernetes
+auth.idam.client.baseUrl=https://idam-api.aat.platform.hmcts.net
+auth.idam.client.redirect-url=https://div-pfe-aat.service.core-compute-aat.internal/authenticated
+idam.s2s-auth.url=http://rpe-service-auth-provider-aat.service.core-compute-aat.internal
+core_case_data.api.url=http://ccd-data-store-api-aat.service.core-compute-aat.internal
+
+auth.idam.client.secret=${idam-secret:replace_me}
+auth.provider.service.client.key=${ccd-submission-s2s-auth-secret:replace_me}
+
+http.proxy=http://proxyout.reform.hmcts.net:8080
```
Comment (hosainnet) [inline src/integrationTest/resources/example-application-local.properties]: For running int tests locally

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/controller/support/GlobalExceptionHandler.java [hunk 8]
```diff
@@ -12,12 +13,21 @@
...
 
-        return ResponseEntity.status(exception.status()).body(
-            String.format("%s - %s", exception.getMessage(), exception.contentUTF8())
+        int statusCode = exception.status();
+
+        String contentUTF8 = exception.contentUTF8();
+        if (contentUTF8 != null && contentUTF8.contains(CASE_REFERENCE_IS_NOT_VALID)) {
+            //This happens when the case is not found in CCD
+            statusCode = HttpStatus.UNAUTHORIZED.value();
+        }
```
Comment (hosainnet) [inline src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/controller/support/GlobalExceptionHandler.java]: Wrap CCD 400 response to 401 instead when case is not found

File: src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/CcdSubmissionITest.java [hunk 9]
```diff
@@ -68,9 +65,6 @@
         (String)ReflectionTestUtils.getField(CcdSubmissionServiceImpl.class,
             "DIVORCE_CASE_SUBMISSION_EVENT_DESCRIPTION");
 
-    @ClassRule
-    public static WireMockClassRule draftStoreServer = new WireMockClassRule(WireMockSpring.options().port(4601));
-
```
Comment (hosainnet) [inline src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/CcdSubmissionITest.java]: Moved to parent

File: src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/CreateDraftServiceITest.java [hunk 10]
```diff
@@ -62,9 +59,6 @@
     @Autowired
     private MockMvc webClient;
 
-    @ClassRule
-    public static WireMockClassRule draftStoreServer = new WireMockClassRule(WireMockSpring.options().port(4601));
-
```
Comment (hosainnet) [inline src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/CreateDraftServiceITest.java]: Moved to parent

File: src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/DeleteDraftServiceITest.java [hunk 11]
```diff
@@ -38,16 +35,13 @@
...
+public class DeleteDraftServiceITest extends MockSupport {
     private static final String API_URL = "/casemaintenance/version/1/drafts";
     private static final String DRAFTS_CONTEXT_PATH = "/drafts";
 
     @Autowired
     private MockMvc webClient;
 
-    @ClassRule
-    public static WireMockClassRule draftStoreServer = new WireMockClassRule(WireMockSpring.options().port(4601));
-
```
Comment (hosainnet) [inline src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/DeleteDraftServiceITest.java]: Moved to parent

File: src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/MockSupport.java [hunk 12]
```diff
@@ -86,7 +85,28 @@
...
+            .extensions(new ConnectionCloseExtension())
+    );
+
+    @ClassRule
+    public static WireMockClassRule caseFormatterServer = new WireMockClassRule(
+        WireMockSpring
+            .options()
+            .port(4011)
+            .extensions(new ConnectionCloseExtension())
+    );
```
Comment (hosainnet) [inline src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/functionaltest/MockSupport.java]: extension should help with flakey functional tests

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java [hunk 13]
```diff
@@ -51,14 +51,14 @@
     public void givenJWTTokenIsNull_whenLinkRespondent_thenReturnBadRequest() {
         Response cmsResponse = linkRespondent(null, "someCaseId", "someLetterHolderId");
 
-        assertEquals(HttpStatus.BAD_REQUEST.value(), cmsResponse.getStatusCode());
+        assertEquals(HttpStatus.UNAUTHORIZED.value(), cmsResponse.getStatusCode());
     }
 
     @Test
     public void givenNoCase_whenLinkRespondent_thenReturnBadRequest() {
```
Comment (alexwakeman) [inline src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java]: Not a biggie but could change name of test to match new expectations (`thenReturnUnauthorized`), for both these ones.

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java [hunk 14]
```diff
@@ -26,6 +37,23 @@ protected IntegrationTest() {
         this.springMethodIntegration = new SpringIntegrationMethodRule();
     }
 
+    @PostConstruct
+    public void init(){
+        if (!Strings.isNullOrEmpty(httpProxy)) {
+            try {
+                URL proxy = new URL(httpProxy);
+                InetAddress.getByName(proxy.getHost()).isReachable(2000); // check proxy connectivity
```
Comment (alexwakeman) [inline src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java]: I noticed the return value of `isReachable` is ignored... not sure if it still works without a conditional:
```
if (InetAddress.getByName(proxy.getHost()).isReachable(2000))
   ...
```

File: src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/controller/support/GlobalExceptionHandlerUTest.java [hunk 15]
```diff
@@ -38,6 +38,24 @@ public void whenHandleBadRequestException_thenReturnUnderLyingError() {
         assertEquals(errorMessage + " - " + errorContent, response.getBody());
     }
 
+    @Test
+    public void givenInvalidCaseIdBadException_whenHandleBadException_thenReturnNotFound() {
```
Comment (alexwakeman) [inline src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/controller/support/GlobalExceptionHandlerUTest.java]: You've called this `givenInvalidCaseIdBadException_whenHandleBadException_thenReturnNotFound` but should be `givenInvalidCaseIdBadException_whenHandleBadException_thenReturnUnauthorized` right?

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java (hunk 14 above)
Comment (hosainnet) [reply src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java]: Yeah it'll throw an exception if there is a network error
From the JavaDocs
```
     * @return a {@code boolean} indicating if the address is reachable.
     * @throws IOException if a network error occurs
```
But perhaps I should also check the boolean returned as well and throw an error if it's false

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java (hunk 13 above)
Comment (hosainnet) [reply src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java]: Done, thanks 👍

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java (hunk 14 above)
Comment (hosainnet) [reply src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java]: Added

File: src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/controller/support/GlobalExceptionHandlerUTest.java (hunk 15 above)
Comment (hosainnet) [reply src/test/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/controller/support/GlobalExceptionHandlerUTest.java]: Done, thanks 👍

**Pull request merged**